experimental/experimental_examples/cpp_api/taskgraph_filtering.cc

```diff
@@ -0,0 +1,288 @@
+/**
+ * @file    taskgraph_filtering.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This example demonstrates basic functionality of task graph APIs.
+ * It reads an array, compresses the data, decompresses, writes data into
+ * a new array, compares the new array data with original data.
+ *
+ *
+ */
+
+#include <iostream>
+
+#include <tiledb/tiledb>
+#include "experimental/tiledb/common/dag/edge/edge.h"
+#include "experimental/tiledb/common/dag/execution/bountiful.h"
+#include "experimental/tiledb/common/dag/execution/throw_catch.h"
+#include "experimental/tiledb/common/dag/graph/taskgraph.h"
+#include "experimental/tiledb/common/dag/nodes/segmented_nodes.h"
+
+#include <bzlib.h>
+
+using namespace tiledb;
+using namespace tiledb::common;
+
```

Comment:
Making this an actual class will clarify a lot of the code below.

Reply:
I didn't fix this as it didn't make much sense for me to replace tuple with a class for this use case. Do we want to enourage users to use a class here? Then it would make sense.